dcc-storage-client/src/main/java/org/icgc/dcc/storage/client/transport/StorageService.java

```diff
@@ -124,41 +124,43 @@ public void downloadPart(DataChannel channel, Part part, String objectId, File o
       public Void doWithRetry(RetryContext ctx) throws IOException {
         log.debug("Download Part URL: {}", part.getUrl());
 
-        final RequestCallback callback = new RequestCallback() {
+        final RequestCallback requestCallback = new RequestCallback() {
 
           @Override
           public void doWithRequest(final ClientHttpRequest request) throws IOException {
-            // request.getHeaders().setContentLength(channel.getlength());
             request.getHeaders().set(HttpHeaders.RANGE, Parts.getHttpRangeValue(part));
           }
         };
 
-        final ResponseExtractor<HttpHeaders> headersExtractor = new ResponseExtractor<HttpHeaders>() {
+        // return md5 calculated by HashingInputStream
+        final ResponseExtractor<String> headersExtractor = new ResponseExtractor<String>() {
 
           @Override
-          public HttpHeaders extractData(ClientHttpResponse response) throws IOException {
+          public String extractData(ClientHttpResponse response) throws IOException {
             try (HashingInputStream his = new HashingInputStream(Hashing.md5(), response.getBody())) {
               channel.readFrom(his);
-              response.getHeaders().set(HttpHeaders.ETAG, his.hash().toString());
-              return response.getHeaders();
+              return his.hash().toString();
             }
           }
         };
 
         try {
           // the actual GET operation
           log.debug("performing GET {}", part.getUrl());
-          HttpHeaders headers =
-              dataTemplate.execute(new URI(part.getUrl()), HttpMethod.GET, callback, headersExtractor);
-          part.setMd5(cleanUpETag(headers.getETag()));
+          val md5 = dataTemplate.execute(new URI(part.getUrl()), HttpMethod.GET, requestCallback, headersExtractor);
+          part.setMd5(md5);
+          if (part.hasFailedChecksum()) {
+            val msg = String.format("Checksum failed for Part# %d: %s", part.getPartNumber(), part.getMd5());
+            throw new Exception(msg);
```

Comment:
use `checkState` instead. Not only does it throw amore informative exception but it reads better and removes a branch:

```java
checkState(!part.hasFailedChecksum(), "Checksum failed for Part# %d: %s", part.getPartNumber(), part.getMd5()); 
```